lib/plugins/openaps.js

```diff
@@ -0,0 +1,488 @@
+        if (!device.mmtune || moments.when.isAfter(device.mmtune.moment)) {
+          device.mmtune = status.mmtune;
+        }
+      }
+    });
+
+    result.lastLoopMoment = result.lastEnacted && result.lastSuggested ? moment.max(result.lastEnacted.moment, result.lastSuggested.moment) : result.lastEnacted && result.lastEnacted.moment || result.lastSuggested && result.lastSuggested.moment;
+
+    result.status = momentsToLoopStatus({
+      enacted: result.lastEnacted && result.lastEnacted.moment
+      , notEnacted: result.lastNotEnacted && result.lastNotEnacted.moment
+      , suggested: result.lastSuggested && result.lastSuggested.moment
+    }, false, recent);
+
+    return result;
+  };
+
+  openaps.getEventTypes = function getEventTypes ( ) {
+    return [
+      {
+        val: 'Temporary Target'
+        , name: 'Temporary Target'
+        , bg: false, insulin: false, carbs: false, prebolus: false, duration: true, percent: false, absolute: false, profile: false, split: false
+        , targets: true, reasons: [
+          { name: 'Eating Soon', targetTop: 80, targetBottom: 80, duration: 60 }
+          , { name: 'Activity', targetTop: 140, targetBottom: 120, duration: 120 }
+          , { name: 'Manual' }
+        ]
+      }
+      , {
+        val: 'Temporary Target Cancel'
+        , name: 'Temporary Target Cancel'
+        , bg: false, insulin: false, carbs: false, prebolus: false, duration: false, percent: false, absolute: false, profile: false, split: false
+      }
+      , {
+        val: 'OpenAPS Offline'
+        , name: 'OpenAPS Offline'
+        , bg: false, insulin: false, carbs: false, prebolus: false, duration: true, percent: false, absolute: false, profile: false, split: false
+      }
+    ];
+  };
+
+  openaps.checkNotifications = function checkNotifications(sbx) {
+    var prefs = openaps.getPrefs(sbx);
+
+    if (!prefs.enableAlerts) { return; }
+
+    var prop = sbx.properties.openaps;
+
+    if (!prop.lastLoopMoment) {
+      console.info('OpenAPS hasn\'t reported a loop yet');
+      return;
+    }
+
+    var now = moment();
+    var level = statusLevel(prop, prefs, sbx);
+    if (level >= levels.WARN) {
+      sbx.notifications.requestNotify({
+        level: level
+        , title: 'OpenAPS isn\'t looping'
+        , message: 'Last Loop: ' + formatAgo(prop.lastLoopMoment, now.valueOf())
+        , pushoverSound: 'echo'
+        , plugin: openaps
+        , debug: prop
+      });
+    }
+  };
+
+  openaps.findOfflineMarker = function findOfflineMarker(sbx) {
+    return _.findLast(sbx.data.treatments, function match(treatment) {
+      var eventTime = sbx.entryMills(treatment);
+      var eventEnd = treatment.duration ? eventTime + times.mins(treatment.duration).msecs : eventTime;
+      return eventTime <= sbx.time && treatment.eventType === 'OpenAPS Offline' && eventEnd >= sbx.time;
+    });
+  };
+
+  openaps.updateVisualisation = function updateVisualisation (sbx) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/2573265186/issues/source?bid=2452374&fileBranchId=3298605#l266)